File: packages/flutter/lib/src/pipeline.dart
import 'dart:async';
import 'api_client.dart';
import 'face_geometry.dart';
import 'face_tracker.dart';
import 'types.dart';

/// Orchestrates detection → tracking → quality selection → API sending.
///
/// Usage:
/// ```dart
/// final session = FaceCaptureSession(
///   apiBaseUrl: 'https://api.example.com',
///   token:      jwtToken,
///   sessionType: 'passive',
///   onSelect: (face) => setState(() => _faces.add(face)),
///   onSessionExpired: () => Navigator.pushNamed(context, '/connect'),
/// );
///
/// // Each camera frame:
/// session.processDetections(detections, frameBuffer);
///
/// // On dispose:
/// session.dispose();
/// ```
class FaceCaptureSession {
  FaceCaptureSession({
    required String apiBaseUrl,
    required String token,
    required this.sessionType,
    this.batchIntervalSeconds = 10,
    this.heartbeatIntervalSeconds = 30,
    this.onSelect,
    this.onTrackRemoved,
    this.onBatchSent,
    this.onSessionExpired,
    this.onCameraNotFound,
  }) : _api = ApiClient(baseUrl: apiBaseUrl, token: token) {
    if (sessionType == 'passive') {
      _batchTimer = Timer.periodic(
        Duration(seconds: batchIntervalSeconds),
        (_) => flushBatch(),
      );
    }
    _heartbeatTimer = Timer.periodic(
      Duration(seconds: heartbeatIntervalSeconds),
      (_) => _heartbeat(),
    );
    _heartbeat(); // immediate first check

    // Tick timer — flush stale tracks
    _tickTimer = Timer.periodic(const Duration(milliseconds: 100), (_) => _tick());
  }

  final String sessionType;
  final int    batchIntervalSeconds;
  final int    heartbeatIntervalSeconds;

  final void Function(SelectedFace face)? onSelect;
  final void Function(String trackId)?   onTrackRemoved;
  final void Function(int count)?        onBatchSent;
  final void Function()?                 onSessionExpired;
  final void Function()?                 onCameraNotFound;

  final ApiClient _api;

  final List<TrackedFace> _tracks = [];
  final List<SelectedFace> _buffer = [];

  Timer? _batchTimer;
  Timer? _heartbeatTimer;
  Timer? _tickTimer;

  int get trackedCount => _tracks.length;
  List<SelectedFace> get pendingFaces => List.unmodifiable(_buffer);

  // ── Public API ──────────────────────────────────────────────────────────────

  /// Feed one frame's ML Kit detections into the tracking pipeline.
  void processDetections(List<Detection> detections, FrameBuffer frame) {
    final valid = nmsFilter(detections);
    updateTracks(valid, frame, _tracks, (t) => commitPending(t, _onSelect));
  }

  /// Send all buffered faces to the API and clear the buffer.
  Future<void> flushBatch() async {
    final nowMs = DateTime.now().millisecondsSinceEpoch;
    final toSend = _buffer.where((f) => f.createdAt.millisecondsSinceEpoch < nowMs - 5000).toList();
    _buffer.removeWhere((f) => f.createdAt.millisecondsSinceEpoch < nowMs - 5000);
    if (toSend.isEmpty) return;
    try {
      await _api.capture(toSend.map((f) => f.dataUrl).toList());
      onBatchSent?.call(toSend.length);
    } catch (_) {
      // Put back on failure
      _buffer.insertAll(0, toSend);
    }
  }

  /// Release all timers. Call from State.dispose() or widget teardown.
  void dispose() {
    _batchTimer?.cancel();
    _heartbeatTimer?.cancel();
    _tickTimer?.cancel();
    _tracks.clear();
    _buffer.clear();
  }

  // ── Private ─────────────────────────────────────────────────────────────────

  void _tick() {
    final updated = flushStale(
      _tracks,
      (t) => commitPending(t, _onSelect),
      (id) => onTrackRemoved?.call(id),
    );
    _tracks
      ..clear()
      ..addAll(updated);
  }

  void _onSelect(SelectedFace face) {
    final idx = _buffer.indexWhere((f) => f.trackId == face.trackId);
    if (idx != -1) {
      _buffer[idx] = face;
    } else {
      _buffer.insert(0, face);
    }
    onSelect?.call(face);
  }

  Future<void> _heartbeat() async {
    try {
      await _api.heartbeat();
    } catch (e) {
      if (e.toString().contains('401')) {
        onSessionExpired?.call();
      } else if (e.toString().contains('404')) {
        onCameraNotFound?.call();
      }
    }
  }
}

File: packages/core/src/pipeline.ts
import { nmsFilter } from "./face-geometry";
import {
  commitPending,
  flushStale,
  updateTracks,
  type TrackedFace,
} from "./face-tracker";
import { ApiClient } from "./api-client";
import type { Detection, FrameBuffer, SelectedFace } from "./types";

export interface FaceCaptureSessionOptions {
  apiBaseUrl: string;
  token: string;
  sessionType: "interactive" | "passive";

  /** How often to batch-send in passive mode (ms). Default: 10 000 */
  batchIntervalMs?: number;
  /** How often to poll /camera/me (ms). Default: 30 000 */
  heartbeatIntervalMs?: number;

  /** Called each time a new best face is selected for a track. */
  onSelect?: (face: SelectedFace) => void;
  /** Called when a track is removed (face left the frame). */
  onTrackRemoved?: (trackId: string) => void;
  /** Called after each successful batch send. */
  onBatchSent?: (count: number) => void;
  /** Called when /camera/me returns 401. */
  onSessionExpired?: () => void;
  /** Called when /camera/me returns 404. */
  onCameraNotFound?: () => void;
}

/**
 * Orchestrates face tracking, quality selection, and API sending.
 * Platform adapters call `processDetections()` each frame and `tick()` each
 * animation frame. Dispose with `dispose()` on unmount.
 */
export class FaceCaptureSession {
  private readonly api: ApiClient;
  private readonly opts: Required<FaceCaptureSessionOptions>;

  private tracks: TrackedFace[] = [];
  private buffer: SelectedFace[] = [];

  private batchTimer: ReturnType<typeof setInterval> | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;

  constructor(opts: FaceCaptureSessionOptions) {
    this.opts = {
      batchIntervalMs:    opts.batchIntervalMs    ?? 10_000,
      heartbeatIntervalMs: opts.heartbeatIntervalMs ?? 30_000,
      onSelect:           opts.onSelect           ?? (() => {}),
      onTrackRemoved:     opts.onTrackRemoved      ?? (() => {}),
      onBatchSent:        opts.onBatchSent         ?? (() => {}),
      onSessionExpired:   opts.onSessionExpired    ?? (() => {}),
      onCameraNotFound:   opts.onCameraNotFound    ?? (() => {}),
      ...opts,
    };

    this.api = new ApiClient(opts.apiBaseUrl, opts.token);

    if (opts.sessionType === "passive") {
      this.batchTimer = setInterval(() => void this.flushBatch(), this.opts.batchIntervalMs);
    }

    this.heartbeatTimer = setInterval(() => void this._heartbeat(), this.opts.heartbeatIntervalMs);
    void this._heartbeat(); // immediate first check
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  get trackedCount(): number { return this.tracks.length; }
  get pendingFaces(): SelectedFace[] { return [...this.buffer]; }
  get activeTracks(): readonly TrackedFace[] { return this.tracks; }

  /**
   * Feed one frame's worth of detections into the pipeline.
   * Call this for every video frame (or every ML Kit result).
   *
   * @param detections  Normalised Detection[] from the platform detector.
   * @param frame       FrameBuffer wrapping the current video frame pixels.
   */
  processDetections(detections: Detection[], frame: FrameBuffer): void {
    const valid = nmsFilter(detections);
    const onCommit = (track: TrackedFace) => commitPending(track, this._onSelect);
    updateTracks(valid, frame, this.tracks, onCommit);
  }

  /**
   * Flush stale tracks. Call periodically (e.g. each rAF or on a timer).
   */
  tick(): void {
    this.tracks = flushStale(
      this.tracks,
      (t) => commitPending(t, this._onSelect),
      (id) => this.opts.onTrackRemoved(id),
    );
  }

  /**
   * Immediately send all buffered faces to the API then clear the buffer.
   * Passive mode calls this automatically on the batch timer; you can also
   * call it manually (e.g. on page hide / app background).
   */
  async flushBatch(): Promise<void> {
    const nowMs = Date.now();
    const faces = this.buffer.filter((f) => f.createdAt < nowMs - 5_000);
    this.buffer = this.buffer.filter((f) => f.createdAt >= nowMs - 5_000);
    if (faces.length === 0) return;
    try {
      await this.api.capture(faces.map((f) => f.dataUrl));
      this.opts.onBatchSent(faces.length);
    } catch {
      // Put them back on failure so we don't lose data
      this.buffer.unshift(...faces);
    }
  }

  /** Stop all timers. Call on component unmount / widget dispose. */
  dispose(): void {
    if (this.batchTimer)     clearInterval(this.batchTimer);
    if (this.heartbeatTimer) clearInterval(this.heartbeatTimer);
    this.tracks = [];
    this.buffer = [];
  }

  // ── Private ────────────────────────────────────────────────────────────────

  private readonly _onSelect = (face: SelectedFace): void => {
    // Update or insert the face slot for this track in the buffer
    const idx = this.buffer.findIndex((f) => f.trackId === face.trackId);
    if (idx !== -1) {
      this.buffer[idx] = face;
    } else {
      this.buffer.unshift(face);
    }
    this.opts.onSelect(face);
  };

  private async _heartbeat(): Promise<void> {
    try {
      await this.api.heartbeat();
    } catch (err) {
      const status = (err as { status?: number }).status;
      if (status === 401) this.opts.onSessionExpired();
      else if (status === 404) this.opts.onCameraNotFound();
    }
  }
}
